src/harness/sas_test_harness.py

```diff
@@ -0,0 +1,364 @@
+      self.end_headers()
+      self.wfile.write(json.load(open(self.__getDumpFilePath('FAD'))))
+    else:
+      url_encoded_id = self.path.split('/')[-1]
+      if not self.__checkDumpFileExist(url_encoded_id):
+        self.send_error(404)
+        return
+      self.send_response(200)
+      self.send_header('Content-type', 'application/json')
+      self.end_headers()
+      self.wfile.write(json.load(open(self.__getDumpFilePath(url_encoded_id))))
+
+# Helper functions for FAD related tests.
+def generateCbsdReferenceId(fcc_id, serial_number):
+  """creates encoded cbsd_reference_id using sha1 with fcc_id and serial number of cbsd.
+  """
+  return str(fcc_id + '/' + str(hashlib.sha1(serial_number).hexdigest())).encode('utf-8')
+
+def generateCbsdRecords(cbsd_records, grant_records_list):
+  """ Generates the cbsdData Object by combining cbsdrecord and grantrecord based on SAS-SAS TS.
+
+  Args:
+     cbsd_records: list of cbsdData objects that includes one or more grants object.
+     grant_records_list: list of muliple grant records for all cbsd devices.
+       For an example : muliple grants [ [grant1, grant2], [grant1, grant2] ] are
+       associated with each cbsd records [ cbsd1, cbsd2 ].
+
+  Returns:
+     It returns list of CBSDData Object based on SAS-SAS TS.
+  """
+
+  cbsd_records_list = []
+  for index, (cbsd_record, grant_records) in enumerate(zip(cbsd_records, grant_records_list)):
+    encoded_cbsd_id = generateCbsdReferenceId(cbsd_record['fccId'],
+                                         cbsd_record['cbsdSerialNumber'])
+
+    # Calculate grant_expiry time for grant objects if grant expiry time may
+    # be incorrect or missing.
+    grant_time_stamp = datetime.now().today() + timedelta(hours=24)
+    grant_expiry_time = time.strftime(grant_time_stamp.replace(microsecond=0).isoformat()) + 'Z'
+
+    internal_data = {
+        'id': 'cbsd/{}'.format(encoded_cbsd_id),
+        'registration': {
+            'fccId': cbsd_record['fccId'],
+            'cbsdCategory': cbsd_record['cbsdCategory'],
+            'callSign': cbsd_record['callSign'],
+            'userId': cbsd_record['userId'],
+            'airInterface': cbsd_record['airInterface'],
+            'measCapability': cbsd_record['measCapability'],
+            'installationParam': cbsd_record['installationParam']
+        }
+    }
+
+    # Check if groupingParam optional parameter present in CBSDRecord then
+    # add to registration object.
+    if cbsd_record.has_key('groupingParam'):
+      internal_data['registration'].update({'groupingParam': cbsd_record['groupingParam']})
+    internal_data_grant = {}
+    internal_data_grant['grants'] = []
+
+    # Read all grant records if mulitiple grant records are associated with single cbsd.
+    for grant_record in grant_records:
+      grant_data = {}
+      if not grant_record.has_key('id'):
+        grant_data['id'] = 'SAMPLE_ID_{:05}'.format(random.randrange(1, 10 ** 5))
```

Comment:
GrantRequest will not have an ID (same for many of the fields below). Pick and input format and work from that.

Reply:
We have   already considered